### PR TITLE
GH-43996: [Java] Mark new allocated ArrowSchema as released

### PR DESCRIPTION
**Pull request by viirya**
### Rationale for this change

As described in #43996.

### Are there any user-facing changes?

* GitHub Issue: #43996

### CODE REVIEW

File: java/c/src/main/java/org/apache/arrow/c/ArrowSchema.java [hunk 1]
```diff
@@ -103,14 +104,21 @@ public static ArrowSchema wrap(long memoryAddress) {
    * @return A new ArrowSchema instance
    */
   public static ArrowSchema allocateNew(BufferAllocator allocator) {
-    return new ArrowSchema(allocator.buffer(ArrowSchema.SIZE_OF));
+    ArrowSchema schema = new ArrowSchema(allocator.buffer(ArrowSchema.SIZE_OF));
+    schema.markReleased();
+    return schema;
```
Comment (lidavidm) [inline java/c/src/main/java/org/apache/arrow/c/ArrowSchema.java:109]: Would it be safer to simply zero out the entire buffer?

**Review by lidavidm — APPROVED** [review]
Could we add a unit test?

File: java/c/src/main/java/org/apache/arrow/c/ArrowSchema.java (hunk 1 above)
Comment (viirya) [reply java/c/src/main/java/org/apache/arrow/c/ArrowSchema.java:109]: I think marking them released should be enough. According to the spec, to use the structures, the users must check if they are not released yet.

**Comment by viirya** [comment]
> Could we add a unit test?

Sure. I was looking for an existing test file to add test for it, but didn't find one for ArrowArray/ArrowSchema. I will find one or create a new one instead.

Thank you.

File: java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java [hunk 2]
```diff
@@ -50,6 +50,14 @@ void afterEach() {
     allocator.close();
   }
 
+  @Test
+  void arraySchemaInit {
+    ArrowSchema schema = ArrowSchema.allocateNew(allocator);
+    Snapshot snapshot = schema.snapshot();
+    assertThat(snapshot.release).isEqualTo(0);
+    schema.close();
+  }
```
Comment (viirya) [inline java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java:59]: Not sure why there is style error in CI. The diff looks good, but in CI it is `voidarraySchemaInit`

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.4.0:check (validate) on project arrow-c-data: Failed during checkstyle configuration: Exception was thrown while processing /build/java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java: IllegalStateException occurred while parsing file /build/java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java. 54:23: no viable alternative at input 'voidarraySchemaInit{': NoViableAltException -> [Help 1]
```

Comment (lidavidm) [reply java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java:59]: It looks wrong here. The parentheses are missing.

**Comment by lidavidm** [comment]
Compilation issue?

```
 Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:testCompile (default-testCompile) on project arrow-c-data: Compilation failure
Error:  /build/java/c/src/test/java/org/apache/arrow/c/ArrowArrayUtilityTest.java:[56,4] error: cannot find symbol
Error:    symbol:   class Snapshot
Error:    location: class ArrowArrayUtilityTest
Error:  -> [Help 1]
```

**Comment by viirya** [comment]
> Compilation issue?

Yea, fixed it now. I cannot run the test locally so previous commits have some issues.

**Review by lidavidm — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @lidavidm

**Pull request merged**